### ISSUE TITLE
.pyc files don't get recreated when they use a macro from a file that was changed

### ISSUE THREAD

**Issue by paultag** [hylang/hy#1324]
It seems like Hy is using the .pyc cache even if the file we're loading has a newer filesystem mtime than the .pyc file -- Python will default to overwriting the pyc.

This winds up biting people who update a script and wind up very confused about why something isn't being invoked :)

(Just bit me on the version on pypi - anyone know if this is being tracked with another issue?)

**Comment by Kodiologist** [comment]
I can't reproduce this with Hy 0.13.0:

    > echo '(print "bye")' > foo.hy
    > hy foo.hy 
    hello
    > hy foo.hy
    hello
    > echo '(print "bye")' > foo.hy 
    > hy foo.hy                     
    byte

**Comment by gilch** [comment]
I'd start looking here https://github.com/hylang/hy/blob/master/hy/importer.py#L57

@paultag, a reliable reproduction guide would be helpful. There was a recent atomation of hyc #1094 #1269.

**Comment by paultag** [comment]
Weird. The above works for me too, on `hy 0.13.0`. The exact issue is pretty tied up to some code, but there's a few macros and imports happening all over the place. I'm not sure what's going on, exactly yet.

I can't get a minimal reproduction at the moment - I'm seeing weird stuff happening, but I did see this happen in the code, and removing the exact .pyc from the cache caused it to recompile the code and start working. I lost a few minutes on that.

Let me try to figure out why it's doing this

**Comment by paultag** [comment]
Oh crap I know why.

I have a macro in a file and I was only changing the macro. I bet that when you have a macro in one file, require it in another, and generate the file off the macro, it will cache the pyc, and modifying the original file won't cause anything using that macro to update.

**Comment by paultag** [comment]
Writing a reproduction now

**Comment by gilch** [comment]
Is this related to #1268?

**Comment by paultag** [comment]
Sorta, this would still be an issue if that was fixed

**Comment by paultag** [comment]
```
(projector) [paultag@nyx:~/x][05:36 PM] ♥  cat api.hy 
(defmacro yeah []
  (import mmm)
  (mmm.generate-function))
(yeah)
(projector) [paultag@nyx:~/x][05:36 PM] ♥  cat mmm.hy 
(defn generate-function [] `(defn yeah [] (print "hi")))
(projector) [paultag@nyx:~/x][05:36 PM] ♥  cat main.py 
import hy
import api
api.yeah()
(projector) [paultag@nyx:~/x][05:37 PM] ♥  python main.py 
hi
(projector) [paultag@nyx:~/x][05:37 PM] ♥  cat mmm.hy 
(defn generate-function [] `(defn yeah [] (print "hi!")))
(projector) [paultag@nyx:~/x][05:37 PM] ♥  python main.py 
hi
(projector) [paultag@nyx:~/x][05:37 PM] ♥  rm -rf __pycache__/
(projector) [paultag@nyx:~/x][05:37 PM] ♥  python main.py 
hi!
```

**Comment by paultag** [comment]
The issue is things we require may change the bytecode, and we have no way of knowing without going through requires and invalidating if the required file(s) have been updated after our bytecode.

I remain skeptical hacking that on top would work. I don't know that we should be writing .pyc files out by default -- maybe we need some sort of production-mode or compile-package helper and let admins do this explicitly?

**Comment by paultag** [comment]
I actually made this example evil by not using require - so that it proves the point about how hard this would be to detect

**Comment by Kodiologist** [comment]
The fact that macros are only run when a file is byte-compiled is a feature, not a bug. As you note, there's no way to check that a macro *would* expand to something different from last time without actually running it.

    $ echo '(defmacro m [] (import random) (random.random)) (print (m))' > foo.hy
    $ foo.hy 
    0.9039111761737755
    $ hy foo.hy
    0.9039111761737755

So I'm inclined to regard this a gotcha that the programmer needs to be aware of, as opposed to a Hy bug.

**Comment by paultag** [comment]
I didn't know .pyc files were being generated - this is a huge gotcha, and I'm going to claim that we ought to not magically write out pyc without a human saying they should

**Comment by gilch** [comment]
> I didn't know .pyc files were being generated

It should absolutely be documented. This was bigger change than we thought.

@Kodiologist is calling this a feature, not a bug.

 I think macros are pointless without compilation. If you're interpreting only, you might as well use [fexprs](https://en.wikipedia.org/wiki/Fexpr), right?

But how do Clojure and Common Lisp handle this kind of situation? We should compare those before deciding. As I recall, some CL implementations have both a compiler and interpreter, but `sbcl`, for example, is only a compiler.

**Comment by Kodiologist** [comment]
> I'm going to claim that we ought to not magically write out pyc without a human saying they should

As of f2278cf2f0297b5223828228311f8ac01c5b36bc, Hy respects Python's environment variable `PYTHONDONTWRITEBYTECODE`. I wouldn't recommend having it the other way around, though, where the programmer has to opt into bytecode. That's forgoing a major speedup that you're going to want 99% of the time. People who really hate bytecode can always put `alias hy='PYTHONDONTWRITEBYTECODE=1 hy'` in their `.bashrc`.

> But how do Clojure and Common Lisp handle this kind of situation?

That's a good question. I was always pretty mystified about how Common Lisp handles these issues.

**Comment by kirbyfan64** [comment]
Can't Hy just track the hashes of both the main file  *and* its requirements?

**Comment by kirbyfan64** [comment]
IMO this is an incredibly bad bug that violates one of the core features of Hy...

**Comment by kirbyfan64** [comment]
Also @paultag is alive :D

**Comment by Kodiologist** [comment]
> Can't Hy just track the hashes of both the main file and its requirements?

No; e.g., https://github.com/hylang/hy/issues/1324#issuecomment-315565440. Similarly, a macro can choose a file to import randomly.

**Comment by gilch** [comment]
> I wouldn't recommend having it the other way around, though, where the programmer has to opt into bytecode. That's forgoing a major speedup that you're going to want 99% of the time. 

We made compilation automatic because Python does it that way. The speedup is considerable and compiling each file by hand seems like a pain. But there might be more nuanced options than opt in or out. In any case, auto-compiling Hy itself seems like a good idea.

> But how do Clojure and Common Lisp handle this kind of situation? 

Common Lisp compiles *forms* rather than files. You can compile the *contents* of a file, but most implementations have no option do this at the command line. The standard way to do it is from the repl with the `COMPILE-FILE` [form](http://www.lispworks.com/documentation/HyperSpec/Body/f_cmp_fi.htm#compile-file). We might want to implement similar functionality in our repl. This would make it easier to recompile things in emacs instead of the shell. You can explicitly recompile an individual file with `hyc` now. That is, instead of deleting `__pycache__/`, you could simply `$ hyc api.hy`.

Python also has a `sys.dont_write_bytecode` [flag](https://docs.python.org/3/library/sys.html#sys.dont_write_bytecode). In Python, you can use this to override the behavior for individual import statements. @Kodiologist did your fix support this flag in Hy too? It probably should.

This way we could turn off auto-compilation on a per-import basis while developing, without turning it off for everything else, and then turn it back on for production. Perhaps a context manager or modified import macro could make this easy to do. This kind of flag would be a dynamic variable in Clojure.

Clojure usually compiles files on load. But it also has an ahead-of-time (AOT) compiler. Certain Java interop features require a file to be compiled AOT. This is also usually done at the repl using the `clojure.core/compile` [form](https://clojure.github.io/clojure/clojure.core-api.html#clojure.core/compile).

**Comment by paultag** [comment]
Correctness should always trump speed

**Comment by Kodiologist** [comment]
> @Kodiologist did your fix support this flag in Hy too?

No, it did not. The environment variable is what's checked.

**Comment by Kodiologist** [comment]
I'm going to relabel this issue, since the compiler is working as designed, although the design is questionable.

**Comment by kirbyfan64** [comment]
> Correctness should always trump speed

*This*. This is Python, not C++. The developer expects full correctness *first*.

> As you note, there's no way to check that a macro would expand to something different from last time without actually running it.

But if the required file changes, it's decently likely that the macro will have changed in some way. So just track the hashes of required files, as well as the hashes of the files that are required by the required files and so forth. This is what C build systems do with headers anyway.

IMO this is a pretty horrible gotcha...

**Comment by josephtoles** [comment]
Would it be appropriate if I added this gotcha to the documentation somewhere? I thought this behavior was a bug until I found this issue.

**Comment by Kodiologist** [comment]
Yes, the tutorial now mentions bytecode, but beyond that, I don't think automatic byte-compilation or its consequences are documented.

**Comment by Kodiologist** [comment]
Update: they are now.

**Comment by gldlcks** [comment]
I steped on this issue too, umm...

**Comment by peaceamongworlds** [comment]
> Update: they are now.

Sorry if I'm missing something, but I don't see any reference to this issue in the docs. Could you point it out?

I agree with @Kodiologist - I don't think that this behavior is a bug. It's just a consequence of that fact that macros are fully expanded out at compile time.

Maybe we should include an additional flag for hy to regenerate all `pyc` files, apart from those in the hy package itself? That would make it somewhat convenient to deal with this issue.

**Comment by Kodiologist** [comment]
[The section "Macros" in tutorial.rst](https://docs.hylang.org/en/stable/tutorial.html#macros) begins with an example of automatic byte-compilation and what that means for macro expansion. I didn't mean to imply that there's an example of the kind of gotcha this issue is about.

**Comment by scauligi** [comment]
I agree that nominally this isn't a bug as long as you understand the compile-time vs run-time modes, but it *is* fairly unintuitive for new users (and even more experienced users -- I've already tripped over this on at least a couple occasions).

One possible solution is to have Hy write additional files to the  `__pycache__` that store compile-time dependency information. E.g., if `app.hy` requires `macros.hy`, then this dependency would be written to `__pycache__/app.cpython-39.hyc`.
Later, after a user edits `macros.hy` and then runs or imports `app.hy`, the Hy importer logic can check the dependencies, see that `macros.hy` has been updated, and re-compile `app.hy` before running it.
